Documentation/design-docs/hosting-layer-apis.md

```diff
@@ -0,0 +1,145 @@
+# Hosting layer APIs
+
+Functionality for advanced hosting scenarios is exposed on the `hostfxr` and `hostpolicy` libraries through C-style APIs.
+
+The `char_t` strings in the below API descriptions are defined based on the platform:
+* Windows     - UTF-16 (2-byte `wchar_t`)
+  * Note that `wchar_t` is defined as a [native type](https://docs.microsoft.com/cpp/build/reference/zc-wchar-t-wchar-t-is-native-type), which is the default in Visual Studio.
+* Unix        - UTF-8  (1-byte `char`)
+
+## Host FXR
+
+All exported functions and function pointers in the `hostfxr` library use the `__cdecl` calling convention on the x86 platform.
+
+## Host Policy
+
+All exported functions and function pointers in the `hostpolicy` library use the `__cdecl` calling convention on the x86 platform.
+
+### .NET Core 1.0+
+
+``` C
+int corehost_load(host_interface_t *init)
+```
+
+Initialize `hostpolicy`. This stores information that will be required to do all the processing necessary to start CoreCLR, but it does not actually do any of that processing.
+* `init` - structure defining how the library should be initialized
+
+If already initalized, this function returns success without reinitializing (`init` is ignored).
+
+``` C
+int corehost_main(const int argc, const char_t* argv[])
+```
+
+Run an application.
+* `argc` / `argv` - command-line arguments
+
+This function does not return until the application completes execution. It will shutdown CoreCLR after the application executes.
+
+``` C
+int corehost_unload()
+```
+
+Uninitialize `hostpolicy`.
+
+### .NET Core 2.1+
+``` C
+int corehost_main_with_output_buffer(
+    const int argc,
+    const char_t *argv[],
+    char_t buffer[],
+    int32_t buffer_size,
+    int32_t *required_buffer_size)
+```
+
+Run a host command and return the output. `corehost_load` should have been called with the `host_command` set on the `host_interface_t`. This function operates in the hosting layer and does not actually run CoreCLR.
+* `argc` / `argv` - command-line arguments
+* `buffer` - buffer to populate with the output (including a null terminator)
+* `buffer_size` - size of `buffer` in `char_t` units
+* `required_buffer_size` - if `buffer` is too small, this will be populated with the minimum required buffer size
```

Comment:
If the buffer is `too small`, will buffer be:

- Empty
- Filled with head/beginning of output
- Filled with tail/end of output

if buffer is not too small, will this be unchanged?

Reply:
Update to include that:
`buffer` is unchanged in the case that the size too small.
`required_buffer_size` is set to 0 in the case that the size is not too small.